bus-mapping/src/circuit_input_builder.rs

```diff
@@ -809,6 +808,210 @@ impl<'a> CircuitInputStateRef<'a> {
     }
 }
 
+/// State and Code Access with "keys/index" used in the access operation.
+#[derive(Debug, PartialEq)]
+enum AccessValue {
+    Account { address: Address },
+    Storage { address: Address, key: Word },
+    Code { address: Address },
+}
+
+/// State Access caused by a transaction or an execution step
+#[derive(Debug, PartialEq)]
+struct Access {
+    step_index: Option<usize>,
+    rw: RW,
+    value: AccessValue,
+}
+
+impl Access {
+    fn new(step_index: Option<usize>, rw: RW, value: AccessValue) -> Self {
+        Self {
+            step_index,
+            rw,
+            value,
+        }
+    }
+}
+
+/// Given a trace and assuming that the first step is a *CALL* kind opcode,
+/// return the result if found.
+fn get_call_result(trace: &[GethExecStep]) -> Option<Word> {
+    let depth = trace[0].depth;
+    trace
+        .iter()
+        .find(|s| s.depth == depth)
+        .map(|s| s.stack.nth_last(0).ok())
+        .flatten()
```

Comment:
I'm not sure this is correct.

Assume we have:
```
SSTORE 1
SSTORE 2
CALL A -> Depth = 1
SSTORE 3
RETURN -> Depth = 0
SSTORE 4
CALL B -> Depth = 1
SSTORE 5
REVERT/STOP -> Depth = 0
SSTORE 6
CALL C -> Depth = 1
SSTORE 7
RETURN -> Depth = 0
SSTORE 8
RETURN -> Return of the TX call.
```

With this trace. if you try to look for the result, you will end up with two of them (as there are two calls at the same depth that return different results).

In theory, I guess that you should pick the first result of the resulting Vec. 
But TBH, I would double check this.

BTW, I think we can use `flat_map()` instead of `map().flatten()`. As we don't have sub-slices. But If we do then ignore the comment as the `flatten` will be required to appear alone.

Reply:
First of all, after reviewing the function I noticed a mistake, it should be like this:
```rust
fn get_call_result(trace: &[GethExecStep]) -> Option<Word> {
    let depth = trace[0].depth;
    trace[1..] // Notice the slicing here
        .iter()
        .find(|s| s.depth == depth)
        .map(|s| s.stack.nth_last(0).ok())
        .flatten()
}
```

I think the trace example you described is not like geth does.  It would be like this:
```
SSTORE 1 (depth = 1)
SSTORE 2 (depth = 1)
CALL A  (depth = 1)
SSTORE 3 (depth = 2)
RETURN (depth = 2)
SSTORE 4 (depth = 1)
CALL B (depth = 1)
SSTORE 5 (depth = 2)
REVERT/STOP (depth = 2)
SSTORE 6 (depth = 1)
CALL C (depth = 1)
SSTORE 7 (depth = 2)
RETURN (depth = 2)
SSTORE (depth = 1)
RETURN -> Return of the TX call.
```
If we run the function at `CALL A`, we get the first step with depth =1, which is `SSTORE 4`, so the function should work as expected.

Notice that
1. `Iterator::find` returns the first element that matches
2. the `map` in this function is an `Option::map`
3. The `flatten` is turning an `Option<Option<T>>` into `Option<T>`.

Reply:
ohhhh. Did not fall into the `find` returning the first one.. 

My bad..